File: HW3/numSum.c
#include <stdio.h>
#include <pthread.h>
#include <stdlib.h>
#include <sys/types.h>

//Mingyuan Li
//Computer system
//HW3
//2/17/2016
//User give a range, program use four thread to calculate the sum

int *list;
int sum;

void *part1(void *param);
void *part2(void *param);
void *part3(void *param);
void *part4(void *param);

int main(int argc, char *argv[])
{
    int i;
    if(argc != 3)
    {
        printf("You need three arguments....\n");
        exit(0);
    }
    int start = atoi(argv[1]);
    int end = atoi(argv[2]);
    int length = end - start + 1;

    list = (int*)malloc(sizeof(int) * length);

    for(int i = 0; i < length - 1; i++)
    {
        list[i] = i + start;
        printf("%d\n", list[i]);
    }
    
    int block = length / 4;
    int size1 = block;
    int size2 = block;
    int size3 = block;
    int size4 = length - (3 * block);

    int start1 = 0;
    int start2 = block;
    int start3 = start2 + block;
    int start4 = start3 + block;

    int ray1[2] = {start1, size1};
    int ray2[2] = {start2, size2};
    int ray3[2] = {start3, size3};
    int ray4[2] = {start4, size4};

    pthread_t t1;
    pthread_t t2;
    pthread_t t3;
    pthread_t t4;

    pthread_create(&t1, 0, part1, ray1);
    pthread_create(&t2, 0, part2, ray2);
    pthread_create(&t3, 0, part3, ray3);
    pthread_create(&t4, 0, part4, ray4);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    pthread_join(t3, NULL);
    pthread_join(t4, NULL);

    printf("Sum is %d\n", sum);
}
void *part1(void *param)
{
    //int *p = (int*)param;
    int start = ((int*)param)[0];
    int size = ((int*)param)[1];

    for(int i = 0; i < size; i++)
    {
        sum += list[i + start];
    }
}
void *part2(void *param)
{
    int start = ((int*)param)[0];
    int size = ((int*)param)[1];

    for(int i = 0; i < size; i++)
    {
        sum += list[i + start];
    }
}
void *part3(void *param)
{
    int start = ((int*)param)[0];
    int size = ((int*) param)[1];

    for(int i = 0; i < size; i++)
    {
        sum += list[i + start];
    }
}
void *part4(void *param)
{
    int start = ((int*) param)[0];
    int size = ((int*) param)[1];

    for(int i = 0; i < size; i++)
    {
        sum += list[i + start];
    }
}

File: HW3/NumSum/src/NumSum/NumSum.java
package NumSum;

/**
 * Created by Mingyuan Li on 2016/2/23.
 * Package name: NumSum.
 * Project name: NumSum.
 */
public class NumSum
{
    private static int sum = 0;
    private static int[] list;
    public static void main(String[] args)
    {
        int start = 0;
        int end = 0;
        if (args.length > 0) {
            try {
                start = Integer.parseInt(args[0]);
                end  = Integer.parseInt(args[1]);
            } catch (NumberFormatException e) {
                System.err.println("Argument" + args[0] + " must be an integer.");
                System.exit(1);
            }
        }
        int length = end - start + 1;
        list = new int[length];
        for (int i = 0; i < length - 1; i++) {
            list[i] = i + start;
            System.out.println(list[i]);
        }

        int block = length / 4;
        int size1 = block;
        int size2 = block;
        int size3 = block;
        int size4 = length - (3 * block);
        int start1 = 0;
        int start2 = block;
        int start3 = start2 + block;
        int start4 = start3 + block;

        Thread t1 = new MyThread(1, start1, size1);
        Thread t2 = new MyThread(2, start2, size2);
        Thread t3 = new MyThread(3, start3, size3);
        Thread t4 = new MyThread(4, start4, size4);

        t1.start();
        t2.start();
        t3.start();
        t4.start();
        try
        {
            t1.join();
            t2.join();
            t3.join();
            t4.join();
        }
        catch (Exception e){
            e.printStackTrace();
        }
        System.out.println("Sum is " + sum);
    }
    static class MyThread extends Thread
    {
        private int id;
        private int start;
        private int size;

        public MyThread(int id, int start, int size)
        {
            this.id = id;
            this.start = start;
            this.size = size;
        }
        public void run()
        {
            for (int i = 0; i < size; i++)
            {
                sum += list[i + start];
            }
        }
    }

}
